Montage fix

wmvanvliet opened the pull request:
This one looks better. It makes some small fixes to your PR.

wmvanvliet committed:
Fixes to montage

- Moved viz code to `viz/montage.py`
- Use `os.path.splitext` for determining file extensions
- Match montage on entire name to avoid ambiguity

wmvanvliet committed:
fix path

dengemann commented:
Ah ok:) we leave the git-fu excercise for your next PR ;-) We normally try to avoid closing PRs for cleanup reasons (otherwise boom).

dengemann commented on mne/montages/montage.py:
```diff
@@ -77,23 +77,25 @@ def read_montage(kind, names=None, path=None, scale=True):
+        kind, ext = montages[0]
+        fname = op.join(path, kind + ext)
+    else:
+        kind, ext = op.splitext(kind)
```
nice!

dengemann commented on mne/viz/montage.py:
```diff
@@ -0,0 +1,30 @@
+                  color=(1.0, 1.0, 1.0), scale_factor=1.5)
+    mlab.text(0.01, 0.01, montage.kind, width=0.4)
+    mlab.view(0, 0)
+    return fig
```
I assume this is the old function in its new correct place?

wmvanvliet replied:
yes, a cut/paste job

dengemann replied:
just wanted to makes sure I miss specific changes.

wmvanvliet replied:
small typo in the docs there

wmvanvliet commented:
Yeah sorry about that. Rebasing proved to be a big pain in this case (somehow it flagged all files as being in conflict). My git-fu is not that strong yet.

dengemann commented on mne/viz/tests/test_montage.py:
```diff
@@ -0,0 +1,19 @@
+import numpy as np
+from mne.montages import read_montage
+
+lacks_mayavi = False
```
I think by conventions we call bool flags like this `has_`. We can test this using git grep in the source of the repository.

git grep 'has_'
git grep 'lacks_'

But probably my bad ;-)

dengemann replied:
edited.

wmvanvliet replied:
actually the rest of the code in `utils.py` uses `skip` as variable name.

dengemann commented:
Apart from that one comment LGTM. I'm currently busy on another branch, I'll merge and test tomorrow. Thanks a lot!

dengemann commented:
> Yeah sorry about that.

Don't worry, git-fu is growing by doing PRs ;-)

dengemann commented:
@wmvanvliet out of curiosity, how do you use the montage so far? Does it work for you?

dengemann commented on mne/montages/montage.py:
```diff
@@ -120,12 +121,11 @@ def read_montage(kind, names=None, path=None, scale=True):
         z = 85. * np.cos(np.deg2rad(theta))
         pos = np.c_[x, y, z]
         names_ = data['f0']
```
@wmvanvliet did you have a chance to play with this montage? I think here was an issue with the orientation, I might be wrong.

wmvanvliet replied:
I've just visualized the one montage. Now that this code merges cleanly with the master branch, I can look into it a bit more.

wmvanvliet committed:
lacks_* --> skip

wmvanvliet committed:
fixed typo in doc

wmvanvliet committed:
Fixed permissions

wmvanvliet committed:
Skip missing sensors

When some sensors are defined in the montage file, but not present in
the info structure, they are silently skipped.

wmvanvliet commented on mne/montages/montage.py:
```diff
@@ -164,7 +164,12 @@ def apply_montage(info, montage):
     if not _contains_ch_type(info, 'eeg'):
         raise ValueError('No eeg channels found')
     for pos, name in zip(montage.pos, montage.names):
-        ch_idx = info['ch_names'].index(name)
```
I strongly feel that `apply_montage` should not fail if a sensor is defined in the montage file but not present in the `info` structure; such sensors should just be skipped. This way, the standard 10_5 montage can for example be applied to many EEG sets that follow the default names.

agramfort replied:
Fine with me

dengemann replied:
I would at least issue an error if none of the montage names is present in the info. Herr I would not use a try / except. Just continue if name is not in info['ch_names'].

wmvanvliet committed:
Throw exception when no sensors match

Also don't use try/catch

wmvanvliet commented:
I'm trying to load a BDF file I've recorded. It doesn't store any electrode positions, but follows the 10-20 naming convention, so it can be a good use case for the montage functionality. Some problems arise:
1. `read_raw_edf` should use the new montage loading functionality instead of requiring a hpts file that specified electrode locations.
2. support for hpts files should be added to `montage.py` to replace the existing functionality in `read_raw_edf` and keep things backwards compatible. The code to do this could be lifted from this function.
3. `fit_sphere_to_headshape` requires three cardinal landmarks: nasion, left auricular, right auricular. Without these, functions like `make_eeg_layout` will not work.

dengemann commented:
> support for hpts files should be added to montage.py to replace the existing functionality in > read_raw_edf and keep things backwards compatible. The code to do this could be lifted from this function.

sounds like that will be another PR ...

> fit_sphere_to_headshape requires three cardinal landmarks: nasion, left auricular, right auricular. Without these, functions like make_eeg_layout will not work.

I think I once replaced the internals, don't remember where. I'm pretty sure you don't need fiducials.... Need to recap ...

dengemann commented:
@wmvanvliet 

I think we need to add the data directory somewhere here to make Travis happy https://github.com/mne-tools/mne-python/blob/master/.travis.yml#L86

dengemann commented:
@wmvanvliet see the environment definitions here for the tests? https://github.com/mne-tools/mne-python/blob/master/.travis.yml#L17

Some of them don't mix well with mayavi. I think we should skip the viz tests for all envs that don't install mayavi. See failing tests.

wmvanvliet commented:
Sorry, I don't follow. What data directory? In `viz/tests/test_montage.py`, the test is skipped if mayavi is not installed.

wmvanvliet commented:
Hrm, what does the `scale` flag do?
https://github.com/wmvanvliet/mne-python/blob/montage-fix/mne/montages/montage.py#L56

agramfort commented:
just add mne/montages to setup.py

can you type the code the way you would like to type it? from raw data to add montage to a plot using topomap?

wmvanvliet commented:
```
import mne

# Load BDF file, specifying standard 10-20 naming convention for electrode
# positions
raw = mne.io.read_raw_edf('semantic_priming_experiment.bdf', n_eeg=32, stim_channel=-1,
                          preload=True, hpts='standard_1020')

# Set bads and reference
raw.info['bads'] = ['T7']
raw, ref = mne.io.set_eeg_reference(raw, ['EXG5', 'EXG6'])

# Bandpass filter
raw.filter(0.5, 15, l_trans_bandwidth=0.2, method='iir')

# Extract epochs for two classes 'related' and 'unrelated'
events = mne.find_events(raw)
picks = mne.pick_types(raw.info, meg=False, eeg=True, exclude='bads')
event_id = dict(related=2, unrelated=3)
epochs = mne.Epochs(raw, events, event_id, tmin=-0.1, tmax=1.0, picks=picks,
                    baseline=(None, 0), reject=dict(eeg=100e-6),
                    preload=True)

# Create ERPs for both classes
related = epochs['related'].average()
unrelated = epochs['unrelated'].average()

# Plot the ERPs
l = mne.layouts.make_eeg_layout(epochs.info)
mne.viz.plot_topo([related, unrelated], l)
```

wmvanvliet committed:
Added montages to setup

agramfort commented:
so montage only gives you channel names?

the first think that comes to mind is that we should have a method of BaseRaw as other data formats will need montage.

raw.set_montage('standard_1020')

or maybe you have a better name in mind.

wmvanvliet commented:
in the example above, the montage gives you electrode positions, based on the channel names defined in the BDF file.

the `montages.apply_montage` function applies the montage to any raw object. We could add it as a method of BaseRaw if we want.

Currently, the `read_raw_edf` function takes a `hpts` parameter to specify electrode positions. The new montage facility should deprecate this, as it's easier and more flexible. We could remove the `hpts` flag and require the user to manually call `apply_montage`. In the example I opted for backwards compatibility and specified the montage name as the `hpts` parameter.

wmvanvliet committed:
fixed mayavi check

wmvanvliet committed:
fix mayavi import

agramfort commented:
> in the example above, the montage gives you electrode positions, based on the channel names defined in the BDF file.
> 
> the montages.apply_montage function applies the montage to any raw object. We could add it as a method of BaseRaw if we want.

any raw object or any container that has an info attribute?

> Currently, the read_raw_edf function takes a hpts parameter to specify electrode positions. The new montage facility should deprecate this, as it's easier and more flexible. We could remove the hpts flag and require the user to manually call apply_montage. In the example I opted for backwards compatibility and specified the montage name as the hpts parameter.

is the semantic of hpts file and the montage really equivalent?
anyone? do most people have hpts files?

dengemann commented:
> raw.set_montage('standard_1020')

or `raw.add_montage` as raw.add_proj

I like this idea.

> any raw object or any container that has an info attribute?

actually yes. Would add it to some mixin. Maybe the channels Mixin.

dengemann commented on mne/utils.py:
```diff
@@ -722,24 +722,30 @@ def dec(*args, **kwargs):
 
 
-def requires_mayavi():
+def requires_mayavi(function):
```
@wmvanvliet I think this change creates trouble

wmvanvliet commented:
> Is the semantic of hpts file and the montage really equivalent?

When loading EDF/BDF files, it is. The function expects a single point for each sensor, plus fiducials. Many EEG labs do not have any device to produce these points and would instead rely on approximate positions provided by the manufacturer of the electrode cap. So many do not have .hpts files.

dengemann commented:
@wmvanvliet I'm a bit swamped at the moment. Feel free to move on. I'll be mergin and finalizing in the next days, probably rather next week. Please feel free to add your name to the files you've edited.

dengemann commented:
Thanks @wmvanvliet ! I'm now finishing this.

wmvanvliet commented:
hope it gets merges soon, this is a very useful feature when dealing with EEG data.

dengemann commented:
We have a blocker. No merge unless the rotation issue is solved ... :(
